CONVENTIONS.md

```diff
@@ -95,7 +95,133 @@ metal3-io](https://github.com/metal3-io/metal3-docs/blob/master/design/bare-meta
+generic `ObjectReference` type; instead, use a more specific type, such as
+`ConfigMapNameReference` or `ConfigMapFileReference` (defined in
+[github.com/openshift/api/config/v1](https://github.com/openshift/api/blob/master/config/v1/types.go)).
+If necessary, define a new type and use it.  Omit the "Ref" suffix in the field
+name.  For example, don't do the following:
+
+```go
+// Example is [...]
+type Example struct {
+	// FrobulatorConfigRef specifies [...].
+	FrobulatorConfigRef corev1.LocalObjectReference `json:"frobulatorConfigRef"`
+
+	// DefabulatorRef specifies [...].
+	DefabulatorRef corev1.LocalObjectReference `json:"defabulatorRef"`
+}
+```
+
+Instead, do the following:
+
+```go
+// Example is [...].
+type Example struct {
+	// frobulatorConfig specifies [...].
+	FrobulatorConfig configv1.ConfigMapNameReference `json:"frobulatorConfig"`
+	
+	// defabulator specifies [...].
+	Defabulator LocalDefabulatorReference `json:"defabulator"`
+}
+
+// LocalDefabulatorReference references a defabulator.
+type LocalDefabulatorReference struct {
+	// name is the metadata.name of the referenced defabulator object.
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+}
+```
+
+Following this convention has the disadvantage that API developers may need to
+define additional types.  However, using custom types has the advantage that the
+types can have context-specific godoc that is more useful to the end-user than
+the generic boilerplate of the generic types.
+
+#### Use Resource Name Rather Than Kind in Object References
+
+Use resource names rather than kinds for object references.  For example, don't
+do the following:
+
+```go
+// DefabulatorReference references a defabulator.
+type DefabulatorReference struct {
+	// APIVersion is the API version of the referent.
+	APIVersion string `json:"apiVersion"`
+	// Kind of the referent.
+	Kind string `json:"kind"`
+	// Namespace of the referent.
+	Namespace string `json:"namespace"`
+	// Name of the referent.
+	Name string `json:"name"`
+}
+```
+
+Instead, do the following:
+
+```go
+// DefabulatorReference references a defabulator [...]
+type DefabulatorReference struct {
+	// group of the referent.
+	// +kubebuilder:validation:Required
+	// +required
+	Group string `json:"group"`
+	// resource of the referent.
+	// +kubebuilder:validation:Required
+	// +required
+	Resource string `json:"resource"`
+	// namespace of the referent.
+	// +kubebuilder:validation:Required
+	// +required
+	Namespace string `json:"namespace"`
+	// name of the referent.
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+}
+```
+
+Following this convention has the disadvantage that it deviates from what users
+may be accustomed to from upstream APIs, but it has the advantage that it avoids
+ambiguity and the need for API consumers to resolve an API version and kind to
+the resource group and name that identify the resource.
```

Comment:
I see the advantage with resources. Where are we today with our APIs? Is that widely implemented?

Reply:
Existing APIs are inconsistent, but newer APIs tend to use resources in references.  

Many older APIs use `corev1.ObjectReference`, which uses kind: I used `git -C ~/src/github.com/openshift/api grep -e corev1.ObjectReference -- '**/types*.go' ':!vendor'` and found **30** occurrences of `corev1.ObjectReference` in the apps, authorization, build, image, machine, security, template, and user APIs.  

A few older APIs also have custom reference types that use kind: [route](https://github.com/openshift/api/blob/30b7580fd50d8d262a9cdcde65dbcb8ebb2f8e32/route/v1/types.go#L132-L133), [unidling](https://github.com/openshift/api/blob/30b7580fd50d8d262a9cdcde65dbcb8ebb2f8e32/unidling/v1alpha1/types.go#L35-L36), [oauth](https://github.com/openshift/api/blob/30b7580fd50d8d262a9cdcde65dbcb8ebb2f8e32/oauth/v1/types.go#L283-L284), and [CVO component overrides](https://github.com/openshift/api/blob/30b7580fd50d8d262a9cdcde65dbcb8ebb2f8e32/config/v1/types_cluster_version.go#L222-L225).  

For newer APIs, we have [`configv1.ObjectReference`](https://github.com/openshift/api/blob/30b7580fd50d8d262a9cdcde65dbcb8ebb2f8e32/config/v1/types_cluster_operator.go#L81-L98), which uses resource and is used in a couple newer APIs: [clusteroperator](https://github.com/openshift/api/blob/30b7580fd50d8d262a9cdcde65dbcb8ebb2f8e32/config/v1/types_cluster_operator.go#L52-L57) and [ingress config](https://github.com/openshift/api/blob/30b7580fd50d8d262a9cdcde65dbcb8ebb2f8e32/config/v1/types_ingress.go#L196-L200).  We also have some instances of custom reference types that use resource: [operator status](https://github.com/openshift/api/blob/30b7580fd50d8d262a9cdcde65dbcb8ebb2f8e32/operator/v1/types.go#L134-L135), [controller leader config in `MasterConfig` in the legacy API](https://github.com/openshift/api/blob/30b7580fd50d8d262a9cdcde65dbcb8ebb2f8e32/legacyconfig/v1/types.go#L1493-L1494), and [`LocalSubjectAccessReview` in the authorization API](https://github.com/openshift/api/blob/30b7580fd50d8d262a9cdcde65dbcb8ebb2f8e32/authorization/v1/types.go#L356-L357).  

Upstream also seems to be moving in the direction of using resource instead of kind in references; see https://github.com/kubernetes/kubernetes/pull/87459, https://github.com/kubernetes/community/pull/5748, https://github.com/kubernetes/community/pull/5973, and https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2336-OwnerReference-resource-field/README.md.  So the OpenShift convention is arguably redundant, but I believe it is still helpful to call out this convention for people who are accustomed to the older upstream convention.